keps/sig-scheduling/4247-queueinghint/README.md

```diff
@@ -0,0 +1,1071 @@
+-->
+
+n/a because the scheduler is only the component to rollout this feature.
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+If `scheduler_pending_pods` metric with `queue: unschedulable` label grows and keeps high number, 
+maybe something goes wrong with QueueingHint and Pods are stuck in the queue.
+
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+No. This feature is the internal feature of the scheduler and no impact left outside the scheduler.
+So, just upgrading it and upgrade->downgrade->upgrade are both the same.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+No.
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+###### How can an operator determine if the feature is in use by workloads?
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+This feature is used during all Pods' scheduling if the feature gate is enabled.
+
+###### How can someone using this feature know that it is working for their instance?
+
+<!--
+For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
+for each individual pod.
+Pick one more of these and delete the rest.
+Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
+and operation of this feature.
+Recall that end users cannot usually observe component logs or access metrics.
+-->
+
+n/a
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+<!--
+This is your opportunity to define what "normal" quality of service looks like
+for a feature.
+
+It's impossible to provide comprehensive guidance, but at the very
+high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99.9% of /health requests per day finish with 200 code
+
+These goals will help you determine what you need to measure (SLIs) in the next
+question.
+-->
+
+n/a
```

Comment:
For beta graduation it would be good to define a reasonable SLO based on the `scheduler_pending_pods` you've used in several questions.

Reply:
Well, it very depends on the cluster. 
Even if there are many unscheduled Pods, that might be not a bug (e.g., all Nodes are full, etc).
So, I'm not sure if we can provide any good number on that metric.

@alculquicondor What do you think?

Reply:
How about a different set of metrics (doesn't have to be explicitly the one I've mentioned above). How about looking at `pod_scheduler_duration` or `pod_scheduling_attempts`, or even a set of values that will allow you to say that the scheduler is stable. Other example: 99% of pods in the queue gets scheduled within x mins.

Reply:
> pod_scheduling_attempts

`schedule_attempts_total` depends on the cluster. If the cluster usually reaches out of capacity, there could be many `unschedulable` results. 

> scheduling_attempt_duration_seconds

The more scheduling constraints the Pods in the cluster, the longer one scheduling cycle can be.
(This enhancement has almost nothing to do with the scheduling cycle anyway)

> Other example: 99% of pods in the queue gets scheduled within x mins.

So, the scheduling latency like `pod_scheduling_sli_duration_seconds` also depends on the cluster. If the cluster usually reaches out of capacity, it could be longer.

---

Well... so, all metrics in the scheduler are here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/metrics/metrics.go

I don't see any metrics that can provide the fixed number (not depending on clusters) which we can tell the scheduling queue is working well. Also, so far, I haven't come up with any ideas that we can add new one to the scheduler to have such SLOs. 

I checked the existing KEP, and the one that is also related to the scheduling queue didn't have an SLO here.
https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3521-pod-scheduling-readiness#what-are-the-reasonable-slos-service-level-objectives-for-the-enhancement

Reply:
We can still have an SLO based on a single scheduling attempt `pod_scheduler_duration`, which shouldn't regress.

And then we can say: in a cluster with at least x pending pods in a given window, `schedule_attempts_total` shouldn't be below y for the window. In other words, we should support scheduling z pods/s. 100 pods/s should be our target for the load tests in k8s.

Reply:
+1 to Aldo

Reply:
For reference, you can use this graph https://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=Scheduler&metricname=LoadSchedulingThroughput&TestName=load

As you can see, we have a throughput of ~150 at the 99 percentile.

Reply:
I put it based on the input from Aldo. PTAL.